File: twidere/src/main/java/org/mariotaku/twidere/activity/support/DataImportActivity.java
package org.mariotaku.twidere.activity.support;

import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;

import org.mariotaku.twidere.R;
import org.mariotaku.twidere.fragment.ProgressDialogFragment;
import org.mariotaku.twidere.fragment.support.DataExportImportTypeSelectorDialogFragment;
import org.mariotaku.twidere.util.DataImportExportUtils;
import org.mariotaku.twidere.util.ThemeUtils;

import java.io.File;
import java.io.IOException;

public class DataImportActivity extends ThemedFragmentActivity implements DataExportImportTypeSelectorDialogFragment.Callback {

    private ImportSettingsTask mImportSettingsTask;
    private OpenImportTypeTask mOpenImportTypeTask;
    private Runnable mResumeFragmentsRunnable;

    @Override
    public int getThemeColor() {
        return ThemeUtils.getThemeColor(this);
    }

    @Override
    public int getThemeResourceId() {
        return ThemeUtils.getNoDisplayThemeResource(this);
    }

    @Override
    public void onCancelled(final DialogFragment df) {
        if (!isFinishing()) {
            finish();
        }
    }

    @Override
    public void onDismissed(final DialogFragment df) {
        if (df instanceof DataExportImportTypeSelectorDialogFragment) {
            finish();
        }
    }

    @Override
    protected void onActivityResult(int requestCode, final int resultCode, final Intent data) {
        switch (requestCode) {
            case REQUEST_PICK_FILE: {
                mResumeFragmentsRunnable = new Runnable() {
                    @Override
                    public void run() {
                        if (resultCode == RESULT_OK) {
                            final String path = data.getData().getPath();
                            if (mOpenImportTypeTask == null || mOpenImportTypeTask.getStatus() != AsyncTask.Status.RUNNING) {
                                mOpenImportTypeTask = new OpenImportTypeTask(DataImportActivity.this, path);
                                mOpenImportTypeTask.execute();
                            }
                        } else {
                            if (!isFinishing()) {
                                finish();
                            }
                        }
                    }
                };
                return;
            }
        }
        super.onActivityResult(requestCode, resultCode, data);
    }


    @Override
    protected void onResumeFragments() {
        super.onResumeFragments();
        if (mResumeFragmentsRunnable != null) {
            mResumeFragmentsRunnable.run();
        }
    }

    @Override
    public void onPositiveButtonClicked(final String path, final int flags) {
        if (path == null || flags == 0) {
            finish();
            return;
        }
        if (mImportSettingsTask == null || mImportSettingsTask.getStatus() != AsyncTask.Status.RUNNING) {
            mImportSettingsTask = new ImportSettingsTask(this, path, flags);
            mImportSettingsTask.execute();
        }
    }

    @Override
    protected void onStart() {
        super.onStart();
        setVisible(true);
    }

    @Override
    protected void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (savedInstanceState == null) {
            final Intent intent = new Intent(this, FileSelectorActivity.class);
            intent.setAction(INTENT_ACTION_PICK_FILE);
            startActivityForResult(intent, REQUEST_PICK_FILE);
        }
    }

    static class ImportSettingsTask extends AsyncTask<Object, Object, Boolean> {
        private static final String FRAGMENT_TAG = "import_settings_dialog";

        private final DataImportActivity mActivity;
        private final String mPath;
        private final int mFlags;

        ImportSettingsTask(final DataImportActivity activity, final String path, final int flags) {
            mActivity = activity;
            mPath = path;
            mFlags = flags;
        }

        @Override
        protected Boolean doInBackground(final Object... params) {
            if (mPath == null) return false;
            final File file = new File(mPath);
            if (!file.isFile()) return false;
            try {
                DataImportExportUtils.importData(mActivity, file, mFlags);
                return true;
            } catch (final IOException e) {
                e.printStackTrace();
            }
            return false;
        }

        @Override
        protected void onPostExecute(final Boolean result) {
            final FragmentManager fm = mActivity.getSupportFragmentManager();
            final Fragment f = fm.findFragmentByTag(FRAGMENT_TAG);
            if (f instanceof DialogFragment) {
                ((DialogFragment) f).dismiss();
            }
            if (result != null && result) {
                mActivity.setResult(RESULT_OK);
            } else {
                mActivity.setResult(RESULT_CANCELED);
            }
            mActivity.finish();
        }

        @Override
        protected void onPreExecute() {
            ProgressDialogFragment.show(mActivity, FRAGMENT_TAG).setCancelable(false);
        }

    }

    static class OpenImportTypeTask extends AsyncTask<Object, Object, Integer> {

        private static final String FRAGMENT_TAG = "read_settings_data_dialog";

        private final DataImportActivity mActivity;
        private final String mPath;

        OpenImportTypeTask(final DataImportActivity activity, final String path) {
            mActivity = activity;
            mPath = path;
        }

        @Override
        protected Integer doInBackground(final Object... params) {
            if (mPath == null) return 0;
            final File file = new File(mPath);
            if (!file.isFile()) return 0;
            try {
                return DataImportExportUtils.getImportedSettingsFlags(file);
            } catch (final IOException e) {
                return 0;
            }
        }

        @Override
        protected void onPostExecute(final Integer flags) {
            final FragmentManager fm = mActivity.getSupportFragmentManager();
            final Fragment f = fm.findFragmentByTag(FRAGMENT_TAG);
            if (f instanceof DialogFragment) {
                ((DialogFragment) f).dismiss();
            }
            final DialogFragment df = new DataExportImportTypeSelectorDialogFragment();
            final Bundle args = new Bundle();
            args.putString(EXTRA_PATH, mPath);
            args.putString(EXTRA_TITLE, mActivity.getString(R.string.import_settings_type_dialog_title));
            if (flags != null) {
                args.putInt(EXTRA_FLAGS, flags);
            } else {
                args.putInt(EXTRA_FLAGS, 0);
            }
            df.setArguments(args);
            df.show(mActivity.getSupportFragmentManager(), "select_import_type");
        }

        @Override
        protected void onPreExecute() {
            ProgressDialogFragment.show(mActivity, FRAGMENT_TAG).setCancelable(false);
        }

    }
}

File: twidere/src/main/kotlin/org/mariotaku/twidere/activity/DataImportActivity.kt
package org.mariotaku.twidere.activity

import android.content.Intent
import android.net.Uri
import android.os.AsyncTask
import android.os.Build
import android.os.Bundle
import android.util.Log
import androidx.documentfile.provider.DocumentFile
import androidx.fragment.app.DialogFragment
import org.mariotaku.ktextension.dismissDialogFragment
import org.mariotaku.twidere.R
import org.mariotaku.twidere.TwidereConstants.*
import org.mariotaku.twidere.fragment.DataExportImportTypeSelectorDialogFragment
import org.mariotaku.twidere.fragment.ProgressDialogFragment
import org.mariotaku.twidere.util.DataImportExportUtils
import java.io.File
import java.io.IOException

class DataImportActivity : BaseActivity(), DataExportImportTypeSelectorDialogFragment.Callback {

    private var importSettingsTask: ImportSettingsTask? = null
    private var openImportTypeTask: OpenImportTypeTask? = null
    private var resumeFragmentsRunnable: Runnable? = null

    override fun onCancelled(df: DialogFragment) {
        if (!isFinishing) {
            finish()
        }
    }

    override fun onDismissed(df: DialogFragment) {
        if (df is DataExportImportTypeSelectorDialogFragment) {
            finish()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        when (requestCode) {
            REQUEST_PICK_FILE -> {
                resumeFragmentsRunnable = Runnable {
                    if (resultCode == RESULT_OK && data != null) {
                        val path = data.data!!
                        if (openImportTypeTask == null || openImportTypeTask!!.status != AsyncTask.Status.RUNNING) {
                            val file = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                                DocumentFile.fromSingleUri(this, path)
                            } else {
                                DocumentFile.fromFile(File(path.path))
                            }
                            openImportTypeTask = OpenImportTypeTask(this@DataImportActivity, file)
                            openImportTypeTask!!.execute()
                        }
                    } else {
                        if (!isFinishing) {
                            finish()
                        }
                    }
                }
                return
            }
        }
        super.onActivityResult(requestCode, resultCode, data)
    }


    override fun onResumeFragments() {
        super.onResumeFragments()
        if (resumeFragmentsRunnable != null) {
            resumeFragmentsRunnable!!.run()
        }
    }

    override fun onPositiveButtonClicked(path: Uri?, flags: Int) {
        if (path == null || flags == 0) {
            finish()
            return
        }
        if (importSettingsTask == null || importSettingsTask!!.status != AsyncTask.Status.RUNNING) {
            importSettingsTask = ImportSettingsTask(this, path, flags)
            importSettingsTask!!.execute()
        }
    }

    override fun onStart() {
        super.onStart()
        setVisible(true)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (savedInstanceState == null) {
            val intent = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                val i = Intent(Intent.ACTION_OPEN_DOCUMENT)
                i.type = "*/*"
                i
            } else {
                val i = Intent(this, FileSelectorActivity::class.java)
                i.action = INTENT_ACTION_PICK_FILE
                i
            }
            startActivityForResult(intent, REQUEST_PICK_FILE)
        }
    }

    internal class ImportSettingsTask(
            private val activity: DataImportActivity,
            private val uri: Uri?,
            private val flags: Int
    ) : AsyncTask<Any, Any, Boolean>() {

        override fun doInBackground(vararg params: Any): Boolean? {
            return try {
                DataImportExportUtils.importData(activity, uri, flags)
                true
            } catch (e: IOException) {
                Log.w(LOGTAG, e)
                false
            }

        }

        override fun onPostExecute(result: Boolean?) {
            activity.executeAfterFragmentResumed { activity ->
                activity.supportFragmentManager.dismissDialogFragment(FRAGMENT_TAG)
            }
            if (result != null && result) {
                activity.setResult(RESULT_OK)
            } else {
                activity.setResult(RESULT_CANCELED)
            }
            activity.finish()
        }

        override fun onPreExecute() {
            activity.executeAfterFragmentResumed {
                ProgressDialogFragment.show(it.supportFragmentManager, FRAGMENT_TAG).isCancelable = false
            }
        }

        companion object {
            private const val FRAGMENT_TAG = "import_settings_dialog"
        }

    }

    internal class OpenImportTypeTask(private val activity: DataImportActivity, private val file: DocumentFile?) : AsyncTask<Any, Any, Int>() {

        override fun doInBackground(vararg params: Any): Int? {
            if (file == null) {
                return 0
            }
            if (!file.isFile) return 0
            return try {
                DataImportExportUtils.getImportedSettingsFlags(activity, file)
            } catch (e: IOException) {
                0
            }

        }

        override fun onPostExecute(flags: Int?) {
            activity.executeAfterFragmentResumed { activity ->
                activity.supportFragmentManager.dismissDialogFragment(FRAGMENT_TAG)
            }
            val df = DataExportImportTypeSelectorDialogFragment()
            val args = Bundle()
            args.putParcelable(EXTRA_PATH, file?.uri)
            args.putString(EXTRA_TITLE, activity.getString(R.string.import_settings_type_dialog_title))
            if (flags != null) {
                args.putInt(EXTRA_FLAGS, flags)
            } else {
                args.putInt(EXTRA_FLAGS, 0)
            }
            df.arguments = args
            df.show(activity.supportFragmentManager, "select_import_type")
        }

        override fun onPreExecute() {
            activity.executeAfterFragmentResumed {
                ProgressDialogFragment.show(it.supportFragmentManager, FRAGMENT_TAG).isCancelable = false
            }
        }

        companion object {

            private const val FRAGMENT_TAG = "read_settings_data_dialog"
        }

    }
}
